### ISSUE TITLE
False positive in reportUnnecessaryComparison

### ISSUE THREAD

**Issue by aDotInTheVoid** [microsoft/pyright#6017]
**Describe the bug**
The following code has a false positive 

**Code or Screenshots**
```python
# pyright: strict
from dataclasses import dataclass


@dataclass
class Nonterminal:
    made_of: list[list["GItem"]]


@dataclass
class Opt:
    inner: "GItem"


GItem = Opt | int


def lmao(c1: Opt, c2: Opt) -> bool:
    return c1 == c2.inner
```

Produces:

```
$ pyright ./bug.py 
/home/alona/tmp/pbug/bug.py
  /home/alona/tmp/pbug/bug.py:19:12 - error: Condition will always evaluate to False since the types "Opt" and "Opt | int" have no overlap (reportUnnecessaryComparison)
1 error, 0 warnings, 0 informations 
```

Which is clearly not true:

```
$ python3
Python 3.11.4 (main, Jun  9 2023, 07:59:55) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from bug import *
>>> lmao(Opt(1), Opt(Opt(1)))
True
>>> 
```

Interestingly, removing `class Nonterminal` makes everything typecheck as expected.

**VS Code extension or command-line**

Pyright CLI, `pyright 1.1.327`, installed with `pip`

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.329, which I just published. It will also be included in a future release of pylance.
